### PR TITLE
Add option to evict keys LRU from the sharded redis tables

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds an experimental `redis_max_memory` flag that bounds the redis memory used per data shard. Note that this only applies to the non-primary shards which store the majority of the task and object metadata. Hence, stuff like client metadata is never evicted.

Since profiling data has a nested structure and cannot be LRU evicted, also make profiling controlled by `collect_profiling_data`, and disable it when `redis_max_memory` is set.

Analysis of redis's approximate LRU eviction algorithm: https://github.com/antirez/redis/blob/a2131f907a752e62c78ea6bb719daf9fe2f91402/src/evict.c#L118

We use `maxmemory_samples 10`. There is also a persisted eviction pool of 16 entries. This effectively gives us 26 tries per eviction to hit a old key (lower bound). Let's assume the most recent 30% of keys are required for stable operation, and we evict at 10000 QPS. Then:
 
```
>>> p_no_bad_eviction = 1 - 0.3**26
0.9999999999999746
>>> p_no_bad_eviction_year = p_not_bad_eviction**(10000*60*60*24*365)
0.9920143099318832
```

So a lower bound on reliability with approx LRU eviction is 99% per year. The actual reliability will be much higher of course since it's unlikely we need even 30% of the metadata, and also the eviction pool is persisted over time.

TODO:
- [x] stress test with long-running Ape-X cluster

## Related issue number

https://github.com/ray-project/ray/issues/3306
https://github.com/ray-project/ray/issues/954
https://github.com/ray-project/ray/issues/3452

### CODE REVIEW

**Review by pcmoritz — APPROVED** [review]
This is working as intended for me.

File: src/ray/ray_config.h [hunk 1]
```diff
@@ -200,7 +200,7 @@ class RayConfig {
       : ray_protocol_version_(0x0000000000000000),
         handler_warning_timeout_ms_(100),
         heartbeat_timeout_milliseconds_(100),
-        num_heartbeats_timeout_(100),
+        num_heartbeats_timeout_(300),
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply src/ray/ray_config.h]: Sounds good. It's possible that some of the tests are currently waiting for the full 10s, in which case that will become really slow. If that's the case and we observe that, then we can configure this parameter specifically in those tests.

File: test/failure_test.py [hunk 2]
```diff
@@ -594,7 +594,7 @@ def test_warning_for_dead_node(ray_start_two_nodes):
     ray.services.all_processes[ray.services.PROCESS_TYPE_RAYLET][0].kill()
 
     # Check that we get warning messages for both raylets.
-    wait_for_errors(ray_constants.REMOVED_NODE_ERROR, 2, timeout=20)
+    wait_for_errors(ray_constants.REMOVED_NODE_ERROR, 2, timeout=40)
```
Comment (robertnishihara) [inline test/failure_test.py]: Ah for this test, can we actually do

`internal_config=json.dumps({"num_heartbeats_timeout": 40})` or something like that?

File: python/ray/rllib/train.py [hunk 3]
```diff
@@ -56,6 +56,12 @@ def create_parser(parser_creator=None):
         default=None,
         type=int,
         help="Emulate multiple cluster nodes for debugging.")
+    parser.add_argument(
+        "--ray-redis-max-memory",
+        default=None,
+        type=int,
+        help="--redis-max-memory to pass to Ray."
+        " This only has an affect in local mode.")
```
Comment (llan-ml) [inline python/ray/rllib/train.py]: Hi @ericl What does "local mode" mean? Does this work in multi-node mode?

Comment (llan-ml) [inline python/ray/rllib/train.py]: Hi @ericl What does "local mode" mean? Does this work in multi-node mode?

**Comment by llan-ml** [comment]
For now the memory flush policy do not support multiple redis shards. I notice that this PR can limit the used memory of each redis shard.

If this PR is merged, does it mean we do not need [redis memory flush](https://ray.readthedocs.io/en/latest/redis-memory-management.html) anymore to some extent?

File: python/ray/rllib/train.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/train.py]: It just means that when using a cluster, you need to pass --redis-max-memory to `ray start` and not train.py

**Comment by ericl** [comment]
@llan-ml that's right, this should supercede redis flushing. I updated the doc page to to remove the old flushing documentation.

File: test/failure_test.py (hunk 2 above)
Comment (ericl) [reply test/failure_test.py]: Done

**Pull request merged**